File: ts/test/entity/ap_i/ApIEntity.test.ts


import Path from 'node:path'
import * as Fs from 'node:fs'

import { test, describe, afterEach } from 'node:test'
import assert from 'node:assert'
import { createLiveTransport } from '../../live-runner'
import { runLiveEntity } from '../../live-entity'


import { PublicApisDatabaseSDK, BaseFeature, stdutil } from '../../..'

import {
  envOverride,
  liveClientOptions,
  liveDelay,
  loadEnvLocal,
  makeCtrl,
  makeMatch,
  makeReqdata,
  makeStepData,
  makeValid,
  maybeSkipControl,
} from '../../utility'


// AFTER the imports on purpose: TypeScript hoists `import` above any
// statement in the emitted CommonJS, so a loader placed above them would
// run only after every imported module had already been evaluated - and
// anything reading process.env at module scope would miss these values.
loadEnvLocal(__dirname + '/../../../.env.local')


describe('ApIEntity', async () => {

  // Per-test live pacing. Delay is read from sdk-test-control.json's
  // `test.live.delayMs`; only sleeps when PUBLIC_APIS_DATABASE_TEST_LIVE=TRUE.
  afterEach(liveDelay('PUBLIC_APIS_DATABASE_TEST_LIVE'))

  test('instance', async () => {
    const testsdk = PublicApisDatabaseSDK.test()
    const ent = testsdk.ApI()
    assert(null != ent)
  })


  test('basic', async (t) => {

    const live = 'TRUE' === process.env.PUBLIC_APIS_DATABASE_TEST_LIVE
    for (const op of ['list', 'load']) {
      if (!live && maybeSkipControl(t, 'entityOp', 'ap_i.' + op, live)) return
    }

    
    const setup = basicSetup()
    if (setup.live) {
      return runLiveEntity(setup, {"active":true,"alias":{"field":{}},"fields":[{"active":true,"name":"avgResponseTime","req":false,"short":"Average response time in milliseconds","type":"`$INTEGER`","index$":0},{"active":true,"format":"uri","name":"baseUrl","req":false,"short":"Base URL of the API","type":"`$STRING`","index$":1},{"active":true,"name":"category","req":false,"short":"Category of the API","type":"`$STRING`","index$":2},{"active":true,"name":"cors","req":false,"short":"Whether CORS is enabled","type":"`$BOOLEAN`","index$":3},{"active":true,"format":"date-time","name":"dateAdded","req":false,"short":"Timestamp when API was added to the database","type":"`$STRING`","index$":4},{"active":true,"name":"description","req":true,"short":"Description of the API functionality","type":"`$STRING`","index$":5},{"active":true,"format":"uri","name":"documentationUrl","req":true,"short":"URL to the API documentation","type":"`$STRING`","index$":6},{"active":true,"name":"endpoints","req":false,"short":"Number of endpoints available","type":"`$INTEGER`","index$":7},{"active":true,"format":"float","name":"errorRate","req":false,"short":"Error rate percentage of the API","type":"`$NUMBER`","index$":8},{"active":true,"name":"healthScore","req":false,"short":"Health score of the API (0-100)","type":"`$INTEGER`","index$":9},{"active":true,"name":"id","req":true,"short":"Unique identifier for the API","type":"`$STRING`","index$":10},{"active":true,"format":"date-time","name":"lastChecked","req":false,"short":"Timestamp of last health check","type":"`$STRING`","index$":11},{"active":true,"name":"name","req":true,"short":"Name of the API","type":"`$STRING`","index$":12},{"active":true,"format":"float","name":"reliability","req":false,"short":"Reliability percentage of the API","type":"`$NUMBER`","index$":13},{"active":true,"name":"tags","req":false,"short":"Tags associated with the API","type":"`$ARRAY`","index$":14}],"id":{"field":"id","name":"id"},"name":"ap_i","op":{"list":{"input":"data","name":"list","points":[{"active":true,"args":{"query":[{"active":true,"kind":"query","name":"category","orig":"category","reqd":false,"type":"`$STRING`","index$":0},{"active":true,"example":50,"kind":"query","name":"limit","orig":"limit","reqd":false,"type":"`$INTEGER`","index$":1},{"active":true,"example":0,"kind":"query","name":"offset","orig":"offset","reqd":false,"type":"`$INTEGER`","index$":2}]},"contract":{"id":"GET /api/list","json":"{\"operationId\":\"listApis\",\"parameters\":[{\"description\":\"Filter APIs by category\",\"in\":\"query\",\"name\":\"category\",\"required\":false,\"schema\":{\"type\":\"string\"}},{\"description\":\"Number of APIs to return\",\"in\":\"query\",\"name\":\"limit\",\"required\":false,\"schema\":{\"default\":50,\"maximum\":100,\"minimum\":1,\"type\":\"integer\"}},{\"description\":\"Offset for pagination\",\"in\":\"query\",\"name\":\"offset\",\"required\":false,\"schema\":{\"default\":0,\"minimum\":0,\"type\":\"integer\"}}],\"protocol\":\"http\",\"responses\":{\"200\":{\"content\":{\"application/json\":{\"schema\":{\"properties\":{\"apis\":{\"items\":{\"properties\":{\"avgResponseTime\":{\"description\":\"Average response time in milliseconds\",\"type\":\"integer\"},\"baseUrl\":{\"description\":\"Base URL of the API\",\"format\":\"uri\",\"type\":\"string\"},\"category\":{\"description\":\"Category of the API\",\"type\":\"string\"},\"cors\":{\"description\":\"Whether CORS is enabled\",\"type\":\"boolean\"},\"dateAdded\":{\"description\":\"Timestamp when API was added to the database\",\"format\":\"date-time\",\"type\":\"string\"},\"description\":{\"description\":\"Description of the API functionality\",\"type\":\"string\"},\"documentationUrl\":{\"description\":\"URL to the API documentation\",\"format\":\"uri\",\"type\":\"string\"},\"endpoints\":{\"description\":\"Number of endpoints available\",\"type\":\"integer\"},\"errorRate\":{\"description\":\"Error rate percentage of the API\",\"format\":\"float\",\"maximum\":100,\"minimum\":0,\"type\":\"number\"},\"healthScore\":{\"description\":\"Health score of the API (0-100)\",\"maximum\":100,\"minimum\":0,\"type\":\"integer\"},\"id\":{\"description\":\"Unique identifier for the API\",\"type\":\"string\"},\"lastChecked\":{\"description\":\"Timestamp of last health check\",\"format\":\"date-time\",\"type\":\"string\"},\"name\":{\"description\":\"Name of the API\",\"type\":\"string\"},\"reliability\":{\"description\":\"Reliability percentage of the API\",\"format\":\"float\",\"maximum\":100,\"minimum\":0,\"type\":\"number\"},\"tags\":{\"description\":\"Tags associated with the API\",\"items\":{\"type\":\"string\"},\"type\":\"array\"}},\"required\":[\"id\",\"name\",\"description\",\"documentationUrl\"],\"type\":\"object\"},\"type\":\"array\"},\"count\":{\"description\":\"Total number of APIs\",\"example\":624,\"type\":\"integer\"}},\"type\":\"object\"}}},\"description\":\"Successfully retrieved list of public APIs\"},\"400\":{\"content\":{\"application/json\":{\"schema\":{\"properties\":{\"code\":{\"description\":\"Error code\",\"type\":\"string\"},\"details\":{\"description\":\"Additional error details\",\"type\":\"string\"},\"error\":{\"description\":\"Error message\",\"type\":\"string\"}},\"required\":[\"error\"],\"type\":\"object\"}}},\"description\":\"Bad request - invalid parameters\"},\"500\":{\"content\":{\"application/json\":{\"schema\":{\"properties\":{\"code\":{\"description\":\"Error code\",\"type\":\"string\"},\"details\":{\"description\":\"Additional error details\",\"type\":\"string\"},\"error\":{\"description\":\"Error message\",\"type\":\"string\"}},\"required\":[\"error\"],\"type\":\"object\"}}},\"description\":\"Internal server error\"}},\"securitySource\":\"unspecified\"}","source":"openapi3","version":1},"kind":"http","method":"GET","orig":"/api/list","segments":[{"lit":"api"},{"lit":"list"}],"select":{"exist":["category","limit","offset"]},"transform":{"req":"`reqdata`","res":"`body.apis`"},"index$":0}],"key$":"list"},"load":{"input":"data","name":"load","points":[{"active":true,"args":{},"contract":{"id":"GET /new","json":"{\"operationId\":\"getNewApiForm\",\"parameters\":[],\"protocol\":\"http\",\"responses\":{\"200\":{\"content\":{\"text/html\":{\"schema\":{\"description\":\"HTML page containing the API submission form\",\"type\":\"string\"}}},\"description\":\"Successfully retrieved the API submission form\"},\"500\":{\"content\":{\"application/json\":{\"schema\":{\"properties\":{\"code\":{\"description\":\"Error code\",\"type\":\"string\"},\"details\":{\"description\":\"Additional error details\",\"type\":\"string\"},\"error\":{\"description\":\"Error message\",\"type\":\"string\"}},\"required\":[\"error\"],\"type\":\"object\"}}},\"description\":\"Internal server error\"}},\"securitySource\":\"unspecified\"}","source":"openapi3","version":1},"kind":"http","method":"GET","orig":"/new","segments":[{"lit":"new"}],"select":{},"transform":{"req":"`reqdata`","res":"`body`"},"index$":0}],"key$":"load"}},"relations":{"ancestors":[]},"key$":"ap_i","name__orig":"ap_i","Name":"ApI","name_":"ap_i","name-":"ap-i","NAME":"AP_I","index$":0}, {"active":true,"entity":"ap_i","key$":"BasicApIFlow","kind":"basic","name":"BasicApIFlow","param":{},"step":[{"active":true,"data":{},"input":{},"match":{},"op":"list","spec":[],"valid":[{"apply":"ItemExists","def":{"ref":"ap_i_ref01"}}],"index$":0},{"active":true,"data":{},"input":{"ref":"ap_i_ref01","srcdatavar":"ap_i_ref01_data","suffix":"_dt0"},"match":{},"op":"load","spec":[],"valid":[{"apply":"TextFieldMark","def":{"mark":"Mark01-ap_i_ref01"}}],"index$":1}]}, 'ApI')
    }
    const client = setup.client
    const struct = setup.struct

    const isempty = struct.isempty
    const select = struct.select

    let ap_i_ref01_data = Object.values(setup.data.existing.ap_i)[0] as any

    // LIST
    const ap_i_ref01_ent = client.ApI()
    const ap_i_ref01_match: any = {}

    const ap_i_ref01_list = (await ap_i_ref01_ent.list(ap_i_ref01_match)).map((e: any) => e.data())


    // LOAD
    const ap_i_ref01_match_dt0: any = {}
    ap_i_ref01_match_dt0.id = ap_i_ref01_data.id
    const ap_i_ref01_data_dt0 = (await ap_i_ref01_ent.load(ap_i_ref01_match_dt0)).data()
    assert(ap_i_ref01_data_dt0.id === ap_i_ref01_data.id)


  })
})



function basicSetup(extra?: any) {
  // TODO: fix test def options
  const options: any = {} // null

  // TODO: needs test utility to resolve path
  const entityDataFile =
    Path.resolve(__dirname, 
      '../../../../.sdk/test/entity/ap_i/ApITestData.json')

  // TODO: file ready util needed?
  const entityDataSource = Fs.readFileSync(entityDataFile).toString('utf8')

  // TODO: need a xlang JSON parse utility in voxgig/struct with better error msgs
  const entityData = JSON.parse(entityDataSource)

  options.entity = entityData.existing

  let client = PublicApisDatabaseSDK.test(options, extra)
  const struct = client.utility().struct
  const merge = struct.merge
  const transform = struct.transform

  let idmap = transform(
    ['ap_i01','ap_i02','ap_i03'],
    {
      '`$PACK`': ['', {
        '`$KEY`': '`$COPY`',
        '`$VAL`': ['`$FORMAT`', 'upper', '`$COPY`']
      }]
    })

  const env = envOverride({
    'PUBLIC_APIS_DATABASE_TEST_AP_I_ENTID': idmap,
    'PUBLIC_APIS_DATABASE_TEST_LIVE': 'FALSE',
    'PUBLIC_APIS_DATABASE_TEST_EXPLAIN': 'FALSE',
  })

  idmap = env['PUBLIC_APIS_DATABASE_TEST_AP_I_ENTID']

  const live = 'TRUE' === env.PUBLIC_APIS_DATABASE_TEST_LIVE

  const transport = createLiveTransport()
  if (live) {
    const rawIds = process.env['PUBLIC_APIS_DATABASE_TEST_AP_I_ENTID']
    idmap = rawIds && rawIds.trim() ? JSON.parse(rawIds) : {}
    if (!idmap || Array.isArray(idmap) || typeof idmap !== 'object') {
      throw new Error('Live ENTID must be a JSON object')
    }
    client = new PublicApisDatabaseSDK(merge([
      // FIRST, so the generated fields below win: sdk-test-control.json's
      // test.client.options adds to the live client, it does not redirect it.
      liveClientOptions(),
      {
      },
      // 'extra || {}', not a bare 'extra': struct.merge returns UNDEFINED when the
      // last entry is undefined, and basicSetup is normally called with no
      // argument at all - so a bare 'extra' silently discarded the apikey
      // and server values above and handed the SDK undefined. Harmless
      // while there was nothing in that object; not harmless now.
      extra || {},
      { system: { fetch: transport.fetch } }
    ]))
  }

  const setup = {
    idmap,
    env,
    options,
    client,
    struct,
    data: entityData,
    explain: 'TRUE' === env.PUBLIC_APIS_DATABASE_TEST_EXPLAIN,
    live,
    transport,
    now: Date.now(),
  }

  return setup
}
  

File: rb/test/ap_i_entity_test.rb
# ApI entity test

require "minitest/autorun"
require "json"
require_relative "../PublicApisDatabase_sdk"
require_relative "runner"

class ApIEntityTest < Minitest::Test
  def test_create_instance
    testsdk = PublicApisDatabaseSDK.test(nil, nil)
    ent = testsdk.ApI(nil)
    assert !ent.nil?
  end

  # Feature #4: the entity stream(action, ...) method runs the op pipeline and
  # returns an Enumerator over result items. With the streaming feature active
  # it yields the feature's incremental output; otherwise it falls back to the
  # materialised list so stream always yields.
  def test_stream
    seed = {
      "entity" => {
        "ap_i" => {
          "s1" => { "id" => "s1" },
          "s2" => { "id" => "s2" },
          "s3" => { "id" => "s3" },
        },
      },
    }

    # Fallback: streaming inactive -> yields the materialised list items.
    base = PublicApisDatabaseSDK.test(seed, nil)
    seen = base.ApI(nil).stream("list", nil, nil).to_a
    assert_equal 3, seen.length

    # Inbound: streaming active -> yields each item from the feature.
    cfg = PublicApisDatabaseConfig.shared_config
    if cfg["feature"].is_a?(Hash) && cfg["feature"].key?("streaming")
      sdk = PublicApisDatabaseSDK.test(seed, { "feature" => { "streaming" => { "active" => true } } })
      got = []
      sdk.ApI(nil).stream("list", nil, nil).each do |item|
        if item.is_a?(Array)
          got.concat(item)
        else
          got << item
        end
      end
      assert_equal 3, got.length
    end
  end

  def test_basic_flow
    setup = ap_i_basic_setup(nil)
    # Per-op sdk-test-control.json skip.
    _live = setup[:live] || false
    ["list", "load"].each do |_op|
      _should_skip, _reason = Runner.is_control_skipped("entityOp", "ap_i." + _op, _live ? "live" : "unit")
      if _should_skip
        skip(_reason || "skipped via sdk-test-control.json")
        return
      end
    end
    # The basic flow consumes synthetic IDs from the fixture. In live mode
    # without an *_ENTID env override, those IDs hit the live API and 4xx.
    if setup[:synthetic_only]
      skip "live entity test uses synthetic IDs from fixture — set PUBLIC_APIS_DATABASE_TEST_AP_I_ENTID JSON to run live"
      return
    end
    client = setup[:client]

    # Bootstrap entity data from existing test data.
    ap_i_ref01_data_raw = Vs.items(Helpers.to_map(
      Vs.getpath(setup[:data], "existing.ap_i")))
    ap_i_ref01_data = nil
    if ap_i_ref01_data_raw.length > 0
      ap_i_ref01_data = Helpers.to_map(ap_i_ref01_data_raw[0][1])
    end

    # LIST
    ap_i_ref01_ent = client.ApI(nil)
    ap_i_ref01_match = {}

    ap_i_ref01_list_result = ap_i_ref01_ent.list(ap_i_ref01_match, nil)
    assert ap_i_ref01_list_result.is_a?(Array)

    # LOAD
    ap_i_ref01_match_dt0 = {
      "id" => ap_i_ref01_data["id"],
    }
    ap_i_ref01_data_dt0_loaded = ap_i_ref01_ent.load(ap_i_ref01_match_dt0, nil)
    ap_i_ref01_data_dt0_load_result = Helpers.to_map(ap_i_ref01_data_dt0_loaded.respond_to?(:data_get) ? ap_i_ref01_data_dt0_loaded.data_get : ap_i_ref01_data_dt0_loaded)
    assert !ap_i_ref01_data_dt0_load_result.nil?
    assert_equal ap_i_ref01_data_dt0_load_result["id"], ap_i_ref01_data["id"]

  end
end

def ap_i_basic_setup(extra)
  Runner.load_env_local

  entity_data_file = File.join(__dir__, "..", "..", ".sdk", "test", "entity", "ap_i", "ApITestData.json")
  entity_data_source = File.read(entity_data_file)
  entity_data = JSON.parse(entity_data_source)

  options = {}
  options["entity"] = entity_data["existing"]

  client = PublicApisDatabaseSDK.test(options, extra)

  # Generate idmap via transform.
  idmap = Vs.transform(
    ["ap_i01", "ap_i02", "ap_i03"],
    {
      "`$PACK`" => ["", {
        "`$KEY`" => "`$COPY`",
        "`$VAL`" => ["`$FORMAT`", "upper", "`$COPY`"],
      }],
    }
  )

  # Detect ENTID env override before envOverride consumes it. When live
  # mode is on without a real override, the basic test runs against synthetic
  # IDs from the fixture and 4xx's. Surface this so the test can skip.
  entid_env_raw = ENV["PUBLIC_APIS_DATABASE_TEST_AP_I_ENTID"]
  idmap_overridden = !entid_env_raw.nil? && entid_env_raw.strip.start_with?("{")

  env = Runner.env_override({
    "PUBLIC_APIS_DATABASE_TEST_AP_I_ENTID" => idmap,
    "PUBLIC_APIS_DATABASE_TEST_LIVE" => "FALSE",
    "PUBLIC_APIS_DATABASE_TEST_EXPLAIN" => "FALSE",
  })

  idmap_resolved = Helpers.to_map(
    env["PUBLIC_APIS_DATABASE_TEST_AP_I_ENTID"])
  if idmap_resolved.nil?
    idmap_resolved = Helpers.to_map(idmap)
  end

  if env["PUBLIC_APIS_DATABASE_TEST_LIVE"] == "TRUE"
    merged_opts = Vs.merge([
      # FIRST, so the generated fields below win: sdk-test-control.json's
      # test.client.options adds to the live client, it does not redirect it.
      Runner.live_client_options,
      {
      },
      extra || {},
    ])
    client = PublicApisDatabaseSDK.new(Helpers.to_map(merged_opts))
  end

  live = env["PUBLIC_APIS_DATABASE_TEST_LIVE"] == "TRUE"
  {
    client: client,
    data: entity_data,
    idmap: idmap_resolved,
    env: env,
    explain: env["PUBLIC_APIS_DATABASE_TEST_EXPLAIN"] == "TRUE",
    live: live,
    synthetic_only: live && !idmap_overridden,
    now: (Time.now.to_f * 1000).to_i,
  }
end
